fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Utils.java

```diff
@@ -906,4 +908,54 @@ public static <K, V> V getValueIfExists(Map<K, V> map, K key) {
 
         return value;
     }
+
+    /**
+     * Recursively resolve the column ref for complicated expressions
+     * Throw exception if that ref cannot be found in the operator expression
+     *
+     * @param ref     column ref
+     * @param factory column factory
+     * @param expr    operator expression
+     * @return all resolved columns
+     */
+    public static List<Pair<Table, Column>> resolveColumnRefRecursive(ColumnRefOperator ref, ColumnRefFactory factory,
+                                                                      OptExpression expr) {
+        // consult the factory
+        Pair<Table, Column> tableAndColumn = factory.getTableAndColumn(ref);
+        if (tableAndColumn != null) {
+            return List.of(tableAndColumn);
+        }
+
+        // When deriving stats, the OptExpression is not exist but only a GroupExpression. We cannot resolve the
+        // ColumnRef from it
+        if (expr == null) {
+            return null;
+        }
+
+        // Consult the projection
+        if (expr.getOp().getProjection() != null) {
```

Comment:
why only consider operator's projection, how about logical scan operators?

Reply:
for `LogicalScanOperator`, we will consult the `colRefToColumnMetaMap` first, at `        PredicateColumnsMgr.getInstance().recordScanColumns(colRefToColumnMetaMap, table, context.getOptExpression())`.
for other operators, the generated expressions are at `projection`, so here choose to consult the `projection` structure.

do you think there're some other structures we need to take care ?